design/scheduler.md

```diff
@@ -0,0 +1,158 @@
+resources, the task will stay in the `Pending` state until it becomes possible
+to schedule it.
+
+When the state of a task reaches `Assigned`, the dispatcher sends this task to
+the assigned node to start the process of executing it.
+
+Each task will only pass through the scheduler once. Once a task is assigned to
+a node, this decision cannot be revisited. See the [task model](task_model.md)
+for more details on task lifecycle.
+
+## Global service tasks
+
+Both replicated and global service tasks pass through the scheduler. For
+replicated tasks, the scheduler needs to decide which node the task should run
+on. For global service tasks, the job of the scheduler is considerably simpler,
+because the global orchestrator creates these tasks with the `NodeID` field
+already set. In this case, the scheduler only has to confirm that the node
+satisfies all the constraints and other filters, and once it does, advance the
+state to `Assigned`.
+
+## Filters
+
+The scheduler needs to run several checks against each candidate node to make
+sure that node is suitable for running the task. At present, this includes the
+following set of checks:
+
+- Confirming the node is in the `Ready` state, as opposed to `Down` or
+  `Disconnected` and availability is `Active`, as opposed to `Pause` or
+  `Drain`
+- Confirming sufficient resource availability
+- Checking that all necessary plugins are installed on the node
+- Checking that user-specified constraints are satisfied
+- Checking that the node has the correct OS and architecture
+- Checking that host ports aren't used by an existing task on the node
+
+This operates through a mechanism called `Pipeline`. `Pipeline` chains together
+filters that perform these checks.
+
+Filters satisfy a simple interface. For simplicity, there is a `SetTask` method
+that lets a task be loaded into the filter and then checked against several
+candidate nodes. The `SetTask` method can do all the processing that only
+depends on the task and not on the node. This approach can save some redundant
+computation and/or allocations. `Filter` also has a `Check` method that tests
+the most-recently-loaded task against a candidate node, and an `Explain` method
+that provides a human-readable explanation of what an unsuccessful result from
+`Check` means. `Explain` is used to produce a message inside the task that
+explains what is preventing it from being scheduled.
+
+## Scheduling algorithm
+
+The current scheduling algorithm works by building a tree of nodes which is
+specific to the service, and attempting to equalize the total number of tasks
+of this service below the branches of the tree at each level. This is done
+subject to constraints, so a node that, for example, doesn't have enough
+resources to accommodate more tasks, will end up with fewer than its peers.
+
+By default, this tree has only one level, and contains all suitable nodes at
+that level. When [placement preferences](topology.md), the tree can be
```

Comment:
the first part of the sentence is unclear - `When placement preferences are specified, ...`?